processor/src/test/java/com/linecorp/decaton/processor/runtime/ProcessorSubscriptionTest.java

```diff
@@ -0,0 +1,130 @@
+
+public class ProcessorSubscriptionTest {
+    /**
+     * A mock consumer which triggers rebalance listener in poll()
+     * like real KafkaConsumer. (refs: KAFKA-6968)
+     */
+    private static class DecatonMockConsumer extends MockConsumer<String, byte[]> {
+        private boolean needRebalance;
+        private ConsumerRebalanceListener listener;
+        private Collection<TopicPartition> previousAssignment;
+
+        private DecatonMockConsumer() {
+            super(OffsetResetStrategy.LATEST);
+        }
+
+        @Override
+        public synchronized void subscribe(Collection<String> topics, ConsumerRebalanceListener listener) {
+            this.listener = listener;
+            previousAssignment = assignment();
+            needRebalance = true;
+            super.subscribe(topics, listener);
+        }
+
+        @Override
+        public synchronized ConsumerRecords<String, byte[]> poll(long timeout) {
+            if (needRebalance) {
+                listener.onPartitionsRevoked(previousAssignment);
+                listener.onPartitionsAssigned(assignment());
+                needRebalance = false;
+            }
+            return super.poll(timeout);
+        }
+    }
+
+    private static ProcessorSubscription subscription(Consumer<String, byte[]> consumer,
+                                                      SubscriptionStateListener listener,
+                                                      TopicPartition tp) {
+        SubscriptionScope scope = new SubscriptionScope(
+                "subscription",
+                tp.topic(),
+                Optional.empty(),
+                ProcessorProperties.builder().build());
+
+        ProcessorSubscription subscription = new ProcessorSubscription(
+                scope,
+                () -> consumer,
+                ProcessorsBuilder.consuming(scope.topic(),
+                                            (byte[] bytes) -> new DecatonTask<>(
+                                                    TaskMetadata.builder().build(), "dummy", bytes))
+                                 .build(null),
+                scope.props(),
+                listener);
+        return subscription;
+    }
+
+    @Test(timeout = 10000L)
+    public void testStateTransition() throws Exception {
+        TopicPartition tp = new TopicPartition("topic", 0);
+        Consumer<String, byte[]> consumer = spy(new DecatonMockConsumer());
+        List<State> states = Collections.synchronizedList(new ArrayList<>());
+
+        CountDownLatch pollLatch = new CountDownLatch(1);
+        doAnswer(inv -> {
+            Object records = inv.callRealMethod();
+            pollLatch.countDown();
+            return records;
+        }).when(consumer).poll(anyLong());
+
+        ProcessorSubscription subscription = subscription(consumer, states::add, tp);
+
+        subscription.start();
+        pollLatch.await();
+
+        assertEquals(Arrays.asList(State.INITIALIZING,
```

Comment:
<img width="530" alt="image" src="https://user-images.githubusercontent.com/2255161/76723808-f19dff00-678b-11ea-98ff-30c4d340940d.png">

same for below

Reply:
Hmm, seems `expected, actual` is the correct order... (My IDE complains IF I flipped current order)

![image](https://user-images.githubusercontent.com/6070684/76725285-0af57a00-6791-11ea-8bf0-b51c61c57397.png)

Reply:
hmm, that's right, nevermind then.